File: leetcode/algorithms/tree/binary-tree-inorder-traversal-iterative.ts
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */


function inorderTraversal(root: TreeNode | null): number[] {
    
    const stack: TreeNode[] = [],trav: number[] = [];
    
    let currNode = root;
    
    while(currNode!==null || stack.length!==0){
        while(currNode!==null){
            stack.push(currNode);
            currNode = currNode.left;
        }
        
        currNode = stack.pop();
        trav.push(currNode.val);
        currNode = currNode.right;
        
    }
    
    
    return trav;
};

File: leetcode/algorithms/tree/binary-tree-inorder-traversal-iterative.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<Integer> inorderTraversal(TreeNode root) {
        
        if(root==null){
            return new ArrayList();
        }
        
        Stack<TreeNode> stk = new Stack<TreeNode>();
        List<Integer> inorderList = new ArrayList<Integer>();
        
        TreeNode curr = root;
        while(curr!=null || !stk.isEmpty()){
            while(curr!=null){
                stk.push(curr);
                curr=curr.left;
            }
            curr = stk.pop();
            inorderList.add(curr.val);
            curr = curr.right;
        }
        
        return inorderList;
    }
}